core/src/program.rs

```diff
@@ -692,23 +714,42 @@ impl<EC: EvalCache> Program<EC> {
 
             for ctr in pending_contracts.iter_mut() {
                 let rt = ctr.contract.clone();
-                ctr.contract = do_eval(vm, rt, current_env.clone(), closurize)?;
+                // Note that contracts can't be referred to recursively, as they aren't binding
+                // anything. Only fields are. This is why we pass `None` for `self_idx`: there is
+                // no locking required here.
+                ctr.contract = eval_guarded(vm, rt, current_env.clone(), closurize)?;
             }
 
             Ok(pending_contracts)
         }
 
-        fn do_eval<EC: EvalCache>(
+        // Handles thunk locking (and unlocking upon errors) to detect infinite recursion, but
+        // hands over the meat of the work to `do_eval`.
+        fn eval_guarded<EC: EvalCache>(
             vm: &mut VirtualMachine<Cache, EC>,
             term: RichTerm,
             env: Environment,
             closurize: bool,
         ) -> Result<RichTerm, Error> {
             vm.reset();
-            let result = vm.eval_closure(Closure {
-                body: term.clone(),
-                env,
-            });
+
+            let curr_thunk = term.as_ref().try_as_closure();
+
+            if let Some(thunk) = curr_thunk.as_ref() {
+                // If the thunk is already locked, it's the thunk of some parent field, and we stop
+                // here to avoid infinite recursion.
+                if !thunk.lock() {
+                    return Ok(term);
```

Comment:
This seems appropriate for docs, but maybe for the deep_seq case we want to raise an error? (If so, it could be in a subsequent PR)

Reply:
Sure, but `deep_seq` isn't using `eval_record_spine` or similar functions anyway, so currently this doesn't affect `deep_seq` in any way. I think it's appropriate for the semantics of `eval_record_spine` in general (doc or not).

Reply:
(Put differently, `deep_seq` impl will handle locking itself and can very much raise an error upon double-lock)